lib/ProductOpener/Display.pm

```diff
@@ -3812,8 +3812,24 @@ sub search_and_display_products($$$$$) {
 	}
 
 	my $count;
+	
+	my $fields_ref;
 
-	my $mongodb_query_ref = [ lc => $lc, query => $query_ref, sort => $sort_ref, limit => $limit, skip => $skip ];
+	#for API (json, xml, rss,...), display all fields
+	if ($request_ref->{json} or $request_ref->{jsonp} or $request_ref->{xml} or $request_ref->{jqm} or $request_ref->{rss}) {
+		$fields_ref = {};
+	} else {
+	#for HTML, limit the fields we retrieve from MongoDB
+		$fields_ref = {
+		"code" => 1,
+		"product_name" => 1,
+		"product_name_$lc" => 1,
+		"brands" => 1,
+		"images" => 1,
+		"quantity" => 1
+		};
+	}
+	my $mongodb_query_ref = [ lc => $lc, query => $query_ref, fields => $fields_ref, sort => $sort_ref, limit => $limit, skip => $skip ];
 
 	my $key = $server_domain . "/" . freeze($mongodb_query_ref);
```

Comment:
I tested it on the .dev server, and I realized that freeze() does not sort hash keys, so at each call we get a different string, which makes the key different almost every time.

Can you replace the key generation with this?

	# tied hashes can't be encoded directly by JSON::PP, freeze the sort tied hash
	my $mongodb_query_ref = [ lc => $lc, query => $query_ref, fields => $fields_ref, sort => freeze($sort_ref), limit => $limit, skip => $skip ];

	# canonical makes JSON::PP sort the keys of hashes
	my $json = JSON::PP->new->utf8->canonical->encode($mongodb_query_ref);

	my $key = $server_domain . "/" . $json;

	$log->debug("MongoDB query key", { key => $key }) if $log->is_debug();

	$key = md5_hex($key);

Reply:
Yes sure I can but you mean you got a different key for same url ? Because I tested on my docker and I get same key for same url.

Reply:
yes, in my case, I got many different keys

the fields were in different orders:

openfoodfacts.dev/12345678
 lcfr query     en:francecountries_tags field▒image▒
product_nam▒product_name_f▒brand▒codquantity sort
                                                 Tie::IxHas▒sortkey sort▒ limi▒ ski▒',
												 
 'openfoodfacts.dev/12345678
 lcfr query     en:francecountries_tags fieldquantit▒cod▒product_name_f▒brand▒
product_nam▒images sort
                       Tie::IxHas▒sortkey sort▒ limi▒ ski▒',request => 'LcAS9CtnFk9ZhRJC'}

Reply:
ok